BoughtTogether/BoughtTogether.java

```diff
@@ -0,0 +1,93 @@
+import java.io.IOException;
+
+import java.util.regex.Pattern;
+import java.util.Arrays;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.IntWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.Mapper;
+import org.apache.hadoop.mapreduce.Reducer;
+import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
+import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
+
+public class BoughtTogether {
+
+    /* generate combinations of items that were bought together */
+    public static class LineMapper
+        extends Mapper<Object, Text, Text, IntWritable> {
+
+        private final static Pattern itemSeparator = Pattern.compile(", *");
+        private final static IntWritable one = new IntWritable(1);
+
+        public void map(Object key, Text value, Context context)
+            throws IOException, InterruptedException {
+
+            /* input lines */
+            String line = value.toString();
+
+            /* split comma-separated items */
+            String[] items = itemSeparator.split(line);
+
+            /* sort the items to prevent duplicates */
+            Arrays.sort(items);
+
+            /* output combinations */
+            for (int i=0; i<items.length; ++i) {
+                String first = items[i];
+                for (int j=i+1; j<items.length; ++j) {
+                    String second = items[j];
+                    Text combination = new Text(first + ", " + second);
```

Comment:
i'll look into fixing that. last time i tried though it added two pairs of parens for some reason. i think i'll ask about the output format in class.